### PR TITLE
Allow using WindowsThread with Mingw

### PR DESCRIPTION
**Pull request by petrutlucian94**
Allow using WindowsThread with Mingw

Most Mingw builds require Posix threads in order to use std::thread.
As per #7764, this is not always the case.

That being considered, we're going to improve the Mingw thread model
checks.

Closes: #7764
Signed-off-by: Lucian Petrut <lpetrut@cloudbasesolutions.com>

### CODE REVIEW

**Comment by mrambacher** [comment]
I am a little confused by the change/code/comment.  There are other places that use/include win_thread.h and uses WindowsThread:
grep win_thread */*/*.{cc,h}
port/win/env_win.cc:#include "port/win/win_thread.h"
port/win/win_thread.cc:#include "port/win/win_thread.h"
port/win/env_win.h:#include "port/win/win_thread.h"
port/win/port_win.h:#include "port/win/win_thread.h"

grep WindowsThread */*/*.{cc,h} | grep -v win_thread
port/win/env_win.cc:    ROCKSDB_NAMESPACE::port::WindowsThread th(&StartThreadWrapper, state.get());
port/win/env_win.h:  std::vector<WindowsThread> threads_to_join_;
port/win/port_win.h:using Thread = WindowsThread;

Won't those places still use WindowsThread, meaning sometimes we would be using posix threads and other times WindowsThread?

**Comment by petrutlucian94** [comment]
> I am a little confused by the change/code/comment. There are other places that use/include win_thread.h and uses WindowsThread:
> grep win_thread _/_/*.{cc,h}
> port/win/env_win.cc:#include "port/win/win_thread.h"
> port/win/win_thread.cc:#include "port/win/win_thread.h"
> port/win/env_win.h:#include "port/win/win_thread.h"
> port/win/port_win.h:#include "port/win/win_thread.h"
> 
> grep WindowsThread _/_/*.{cc,h} | grep -v win_thread
> port/win/env_win.cc: ROCKSDB_NAMESPACE::port::WindowsThread th(&StartThreadWrapper, state.get());
> port/win/env_win.h: std::vector threads_to_join_;
> port/win/port_win.h:using Thread = WindowsThread;
> 
> Won't those places still use WindowsThread, meaning sometimes we would be using posix threads and other times WindowsThread?

ohh, you're actually right. I was relying on the fact that the thread class is picked here https://github.com/facebook/rocksdb/blob/v6.17.3/port/win/port_win.h#L223-L228, but here it's explicitly using WindowsThread: https://github.com/facebook/rocksdb/blob/v6.17.3/port/win/env_win.h#L73

I'll be completely honest, we thought we needed rocksdb for the Windows Ceph port and added https://github.com/facebook/rocksdb/commit/172adce767c87898b55eca7f47f53cfe98c13638 in order to get it to build using mingw. We were using a static build, so apparently not all the required bits were linked.

Since that was an incomplete fix that's now causing issues, we can either revert it completely or consider using ``Thread`` instead of ``WindowsThread`` here:  https://github.com/facebook/rocksdb/blob/v6.17.3/port/win/env_win.h#L73.

**Comment by mrambacher** [comment]
> > I am a little confused by the change/code/comment. There are other places that use/include win_thread.h and uses WindowsThread:
> > grep win_thread _/_/*.{cc,h}
> > port/win/env_win.cc:#include "port/win/win_thread.h"
> > port/win/win_thread.cc:#include "port/win/win_thread.h"
> > port/win/env_win.h:#include "port/win/win_thread.h"
> > port/win/port_win.h:#include "port/win/win_thread.h"
> > grep WindowsThread _/_/*.{cc,h} | grep -v win_thread
> > port/win/env_win.cc: ROCKSDB_NAMESPACE::port::WindowsThread th(&StartThreadWrapper, state.get());
> > port/win/env_win.h: std::vector threads_to_join_;
> > port/win/port_win.h:using Thread = WindowsThread;
> > Won't those places still use WindowsThread, meaning sometimes we would be using posix threads and other times WindowsThread?
> 
> ohh, you're actually right. I was relying on the fact that the thread class is picked here https://github.com/facebook/rocksdb/blob/v6.17.3/port/win/port_win.h#L223-L228, but here it's explicitly using WindowsThread: https://github.com/facebook/rocksdb/blob/v6.17.3/port/win/env_win.h#L73
> 
> I'll be completely honest, we thought we needed rocksdb for the Windows Ceph port and added [172adce](https://github.com/facebook/rocksdb/commit/172adce767c87898b55eca7f47f53cfe98c13638) in order to get it to build using mingw. We were using a static build, so apparently not all the required bits were linked.
> 
> Since that was an incomplete fix that's now causing issues, we can either revert it completely or consider using `Thread` instead of `WindowsThread` here: https://github.com/facebook/rocksdb/blob/v6.17.3/port/win/env_win.h#L73.

I would suggest doing the following:
- Only defining the WindowsThread class if _POSIX_THREAD is not set
- Changing the places that use WindowsThread to use Thread

And see what happens... :-)

**Review by mrambacher — APPROVED** [review]
LGTM.

Could you also make win_thread.h only define WindowsThread if _POSIX_THREADS is not defined?  That would allow quick/easy detection of any use of the wrong thread class.

**Comment by jay-zhuang** [comment]
@petrutlucian94 are we still blocked to setup a CircleCI job to run the MinGW build on Windows? Or are we blocked by chocolatey package which is missing posix thread?

**Comment by petrutlucian94** [comment]
> @petrutlucian94 are we still blocked to setup a CircleCI job to run the MinGW build on Windows? Or are we blocked by chocolatey package which is missing posix thread?

I think setting up a MinGW CI is doable, either on Windows or Linux. As mentioned here https://github.com/facebook/rocksdb/pull/6865#issuecomment-805730931, that MinGW archive can be used instead of the chocolatey package.

Unfortunately my schedule is already tight at the moment, so I can't volunteer.

File: port/win/win_thread.cc [hunk 1]
```diff
@@ -180,4 +185,5 @@ unsigned int __stdcall  WindowsThread::Data::ThreadProc(void* arg) {
 } // namespace port
 }  // namespace ROCKSDB_NAMESPACE
 
-#endif
+#endif  // OS_WIN
+#endif  // !_POSIX_THREADS
```
Comment (cculianu) [inline port/win/win_thread.cc:189]: These two comments are in the wrong order. :)

**Pull request closed without merging**